### PR TITLE
SettingsLog: fix text color bug on theme switch; new warning/green text colors

### LINKED ISSUE #3475: Theme Switch Bug
_**Settings**_ 🠖 **_Log_**: The text of the Daemon log does not change color when changing from Light to Dark Theme and vice versa.

image

### PR DESCRIPTION
**Pull request by rating89us**
Closes #3475

Changes:
- Fix text color bug on theme switch (#3475)
- New warning text color (orange darker than previous orange, for better contrast on light theme)
- New green text color (dark green, for better contrast on light theme)

### CODE REVIEW

**Comment by selsta** [comment]
> New green text color (dark green, for better contrast on light theme)

This has noticeably worse contrast in dark theme.

File: pages/settings/SettingsLog.qml [hunk 1]
```diff
@@ -132,6 +132,20 @@ Rectangle {
...
+                var flickableContentYBefore = flickable.contentY
+                var daemonLogText = consoleArea.text
+                consoleArea.clear();
+                if (MoneroComponents.Style.blackTheme) {
+                    var daemonLogTextSwitchedTextColor = daemonLogText.replace(/#000000/g, '#ffffff');
+                } else {
+                    var daemonLogTextSwitchedTextColor = daemonLogText.replace(/#ffffff/g, '#000000');
+                }
+                consoleArea.append(daemonLogTextSwitchedTextColor);
+                flickable.contentY = flickableContentYBefore
```
Comment (selsta) [inline pages/settings/SettingsLog.qml:147]:
```suggestion
                if (MoneroComponents.Style.blackTheme) {                       
                    consoleArea.text = consoleArea.text.replace(/#000000/g, '#ffffff');   
                } else {                                                            
                    consoleArea.text = consoleArea.text.replace(/#ffffff/g, '#000000');   
                }                                                              
```

Comment (rating89us) [reply pages/settings/SettingsLog.qml:147]: I tried this approach before, but it doesn't work. After switching theme, text area scrolls up (scrollbar position doesn't persist) and every time you type a command (like 'status'), it scrolls up again. That's why I used .clean() and then .append() on consoleArea and reconfigured    flickable.contentY.

Comment (selsta) [reply pages/settings/SettingsLog.qml:147]: Oh, I didn't test this. Will recheck.

**Comment by rating89us** [comment]
> This has noticeably worse contrast in dark theme. Maybe two different colors depending on the theme?

fixed

File: pages/settings/SettingsLog.qml (hunk 1 above)
Comment (rating89us) [reply pages/settings/SettingsLog.qml:147]: this fixes the scroll bar, but this bug is still present: every time you type a command (like 'status'), it scrolls up again.

Comment (rating89us) [reply pages/settings/SettingsLog.qml:147]: read this
https://doc.qt.io/qt-5/qml-qtquick-textedit.html#text-prop

> "The append(), insert() and remove() methods provide more fine-grained control and remarkably better performance for modifying especially large rich text content."

Comment (selsta) [reply pages/settings/SettingsLog.qml:147]: This code is in `onColorChanged` which will rarely get called so performance does not matter here. Apart from that append, insert, remove being faster in not valid if you clear it first.

Comment (selsta) [reply pages/settings/SettingsLog.qml:147]: > every time you type a command (like 'status'), it scrolls up again.

It does not for me. Did you test my exact code?

loop

Comment (rating89us) [reply pages/settings/SettingsLog.qml:147]: Yes, I used your code. Did you switch theme before typing status?

Comment (rating89us) [reply pages/settings/SettingsLog.qml:147]: logtext
This is what happens when I type status, after switching theme.

Comment (selsta) [reply pages/settings/SettingsLog.qml:147]: Yes, can reproduce now. Don't know a better solution than your code.

File: pages/settings/SettingsLog.qml [hunk 2]
```diff
@@ -132,6 +132,20 @@ Rectangle {
...
             font.family: MoneroComponents.Style.fontRegular.name
             text: qsTr("Daemon log") + translationManager.emptyString
+            themeTransition: false
+            onColorChanged: {
+                console.log("color changed")
+                var flickableContentYBefore = flickable.contentY
+                var daemonLogText = consoleArea.text
+                consoleArea.clear();
+                if (MoneroComponents.Style.blackTheme) {
+                    var daemonLogTextSwitchedTextColor = daemonLogText.replace(/#000000/g, '#ffffff');
```
Comment (selsta) [inline pages/settings/SettingsLog.qml:142]: If we want to do it perfectly we would also have to fix the different lime / green color on change.

File: pages/settings/SettingsLog.qml [hunk 3]
```diff
@@ -164,7 +178,11 @@ Rectangle {
                     wrapMode: TextEdit.Wrap
                     readOnly: true
                     function logCommand(msg){
-                        msg = log_color(msg, "lime");
```
Comment (selsta) [inline pages/settings/SettingsLog.qml:167]: Ternary operator would be more readable here IMO and more compact, but that's just personal style.

```
msg = log_color(msg, MoneroComponents.Style.blackTheme ? "lime" : "green")
consoleArea.append(msg);
```

or even in one line but that's pushing readability :D

```
consoleArea.append(log_color(msg, MoneroComponents.Style.blackTheme ? "lime" : "green"));
```

File: pages/settings/SettingsLog.qml (hunk 1 above)
Comment (rating89us) [reply pages/settings/SettingsLog.qml]: yeah, I spent a long time trying to fix this, but this was the only solution I found.

**Review by selsta — APPROVED** [review]

**Pull request merged**